### PR TITLE
Fix #5, use path_max instead of os_max_file_name

### LINKED ISSUE #5: elf2cfetbl internal filename/path length dependency on OSAL questionable
Originated by abrown4 (68 on babelfish):

The internal filename/path length for the table files produced do NOT have to depend upon the OSAL. The OSAL OS_MAX_FILE_NAME limit is too restrictive. Need to use the PATH_MAX from linux/limits.h that is appropriate for the host.

JH - Suggest possibly checking size against OSAL and notify user when limit exceeded (but still allow to run as long as it meets PATH_MAX).

### PR DESCRIPTION
**Pull request by avan989**
**Describe the contribution**
Fix #5, used path_max instead of os_max_file_name. 

**Testing performed**
Steps taken to test the contribution:
1. Build 
2. Create sample_table.tbl file using elf2cfetbl. Used full path to trigger warning:
3. Verify warning and that tbl file is still created. 
```
Lenght of Source Filename + Path exceed OS_MAX_FILE_NAME
Lenght of Destination Filename + Path exceed OS_MAX_FILE_NAME
```

4. Used sample_table.tbl file and verify it still work in cFS. 

**System(s) tested on:**
 - Hardware
 - 18.04.03
 - cFE 6.7.0, elf2cfetbl 3.1.1

**Contributor Info**
Anh Van, NASA Goddard

**Community contributors**
You must attach a signed CLA (required for acceptance) or reference one already submitted

### CODE REVIEW

**Comment by skliper** [comment]
> A couple issues with this.
> 
> I disagree about checking against `OS_MAX_FILE_NAME` for two big reasons.
> 
> 1. This is a full path name on the HOST machine.  The location of the file on the HOST can be very long and this has absolutely no relationship to its location once copied to a target.  So warning if the HOST path exceeds OS_MAX_FILE_NAME doesn't add anything useful, it's just noise.

Disagree.  Line in build is:

`cd /home/jhageman/cFS/cFS-GitHub/build/cpu1/apps/sch_lab/tables_cpu1 && /home/jhageman/cFS/cFS-GitHub/build/tools/elf2cfetbl/elf2cfetbl sch_lab_table.o`

This should warn if the name is too long, since the expectation when is this table should be loadable without any changes.

> 2. This tool should not know what the `osconfig.h` of the target is, at all.  We need to _reduce_ dependencies on target-specific config here (I am going to file a separate bug about the inclusion of cfe_tbl_internal.h which should _not be required after the CFE pull request 27)

It needs to build tables that work with cFE, hence it's by design dependent on cFE settings.  Feel free to suggest alternatives, but this warning has purpose.

> Also, I believe POSIX specifies that `PATH_MAX` is defined in limits.h. So one should `#include <limits.h>` not `#include <linux/limits.h>`. The former is POSIX-compliant but the latter, of course, will only work on Linux.

Agreed.

**Comment by jphickey** [comment]
> > This is a full path name on the HOST machine.  The location of the file on the HOST can be very long and this has absolutely no relationship to its location once copied to a target.  So warning if the HOST path exceeds OS_MAX_FILE_NAME doesn't add anything useful, it's just noise.
> 
> Disagree. Line in build is:
> 
> `cd /home/jhageman/cFS/cFS-GitHub/build/cpu1/apps/sch_lab/tables_cpu1 && /home/jhageman/cFS/cFS-GitHub/build/tools/elf2cfetbl/elf2cfetbl sch_lab_table.o`
> 
> This should warn if the name is too long, since the expectation when is this table should be loadable without any changes.

That is only an artifact of how the build system happens to currently invoke it as part of the current/default build process, because it "cd'd" into the directory first and then ran the tool so it therefore didn't need to specify any relative path to the file.... If you run this from the command line or otherwise write your own Makefile to generate a table (as many do), or even if the CMake recipe changes in the future then this can be a full path name to the source file.  

That particular method of invocation is not a requirement and not guaranteed to remain consistent in the future.

I thought this was the whole reason for changing from OS_MAX_FILE_NAME to PATH_MAX -- because it is a host-side tool and this is storing a host-side path name, so OS_MAX_FILE_NAME is irrelevant here.


> > This tool should not know what the `osconfig.h` of the target is, at all.  We need to _reduce_ dependencies on target-specific config here (I am going to file a separate bug about the inclusion of cfe_tbl_internal.h which should _not be required after the CFE pull request 27)
> 
> It needs to build tables that work with cFE, hence it's by design dependent on cFE settings. Feel free to suggest alternatives, but this warning has purpose.

But CFE doesn't have a definition for this, OSAL does.  And it can be different on every CPU, and elf2cfetbl doesn't know which CPU it is actually generating a table for.

**Comment by skliper** [comment]
> That particular method of invocation is not a requirement and not guaranteed to remain consistent in the future.
...
> I thought this was the whole reason for changing from OS_MAX_FILE_NAME to PATH_MAX -- because it is a host-side tool and this is storing a host-side path name, so OS_MAX_FILE_NAME is irrelevant here.
...
> But CFE doesn't have a definition for this, OSAL does. And it can be different on every CPU, and elf2cfetbl doesn't know which CPU it is actually generating a table for.

Understood it's not perfect, and if build system or invocation changes it may be a false warning but in our current context it's useful for us and not an error.

**Comment by jphickey** [comment]
It is more than just not perfect, it's **not correct**.  Perhaps I didn't express that clearly enough...

- It is validating/checking something that happens to work today only by mere coincidence.    
- It is checking both the source and destination files, where only the destination file name length  even remotely matters (because this is the one that gets put on the target, **not the source file**).  
- It is checking the entire path name against OS_MAX_FILE_NAME -- which is wrong -- this is a path name so it should be OS_MAX_PATH_NAME (but not, because its a host path name)
- The actual file name part (the one that would be validated against OS_MAX_FILE_NAME) is **ALREADY** guaranteed to be less than the maximum size, because it is stored in the `TblFileDef` object in a limited-size buffer.

In other words, the validation you are looking for (that the actual filename part is less than the size limit) is already done, and this extra check is both superfluous and wrong.

To make it even more problematic, this will start to trigger a compiler error once merged with the other cleanup, when you (rightfully) cannot include osconfig.h anymore, because the hacks are taken out.

**Comment by skliper** [comment]
> * The actual file name part (the one that would be validated against OS_MAX_FILE_NAME) is **ALREADY** guaranteed to be less than the maximum size, because it is stored in the `TblFileDef` object in a limited-size buffer.

Good point.  I was thrown off by the destination directory command line option, which just gets appended to.

> In other words, the validation you are looking for (that the actual filename part is less than the size limit) is already done...

And that's what I needed to hear!  I'm on board with just taking out the warnings now.

**Comment by avan989** [comment]
removed warning, osconfig.h and changed to linux.h. Tested with sample_table, still work.

**Comment by avan989** [comment]
update to path_max -1
